dbms/src/Flash/Planner/Plans/PhysicalExpand.cpp

```diff
@@ -0,0 +1,132 @@
+    ExpressionActionsPtr expand_actions_itself = PhysicalPlanHelper::newActions(child->getSampleBlock());
+
+    auto grouping_sets = analyzer.buildExpandGroupingColumns(expand, before_expand_actions);
+    auto expand_action = ExpressionAction::expandSource(grouping_sets);
+    expand_actions_itself->add(expand_action);
+
+    // construct sample block.
+    NamesAndTypes expand_output_columns;
+    auto child_header = child->getSchema();
+    for (const auto & one : child_header)
+    {
+        expand_output_columns.emplace_back(one.name, expand_action.expand->isInGroupSetColumn(one.name) ? makeNullable(one.type) : one.type);
+    }
+    expand_output_columns.emplace_back(expand_action.expand->grouping_identifier_column_name, expand_action.expand->grouping_identifier_column_type);
+
+    auto physical_expand = std::make_shared<PhysicalExpand>(
+        executor_id,
+        expand_output_columns,
+        log->identifier(),
+        child,
+        expand_action.expand,
+        expand_actions_itself,
+        Block(expand_output_columns));
+
+    return physical_expand;
+}
+
+
+void PhysicalExpand::expandTransform(DAGPipeline & child_pipeline)
+{
+    String expand_extra_info = fmt::format("expand, expand_executor_id = {}", execId());
+    FmtBuffer fb;
+    fb.append(": grouping set ");
+    shared_expand->getGroupingSetsDes(fb);
+    expand_extra_info.append(fb.toString());
+    child_pipeline.transform([&](auto & stream) {
+        stream = std::make_shared<ExpressionBlockInputStream>(stream, expand_actions, log->identifier());
+        stream->setExtraInfo(expand_extra_info);
+    });
+}
+
+void PhysicalExpand::buildPipelineExec(PipelineExecGroupBuilder & group_builder, Context &, size_t)
+{
+    auto input_header = group_builder.getCurrentHeader();
+    group_builder.transform([&](auto & builder) {
+        builder.appendTransformOp(std::make_unique<ExpressionTransformOp>(group_builder.exec_status, expand_actions, log->identifier()));
+    });
+}
+
+void PhysicalExpand::buildBlockInputStreamImpl(DAGPipeline & pipeline, Context & context, size_t max_streams)
+{
+    child->buildBlockInputStream(pipeline, context, max_streams);
+    expandTransform(pipeline);
+}
+
+void PhysicalExpand::finalize(const Names & parent_require)
+{
+    FinalizeHelper::checkSchemaContainsParentRequire(schema, parent_require);
+    Names required_output;
+    required_output.reserve(shared_expand->getGroupSetNum()); // grouping set column should be existed in the child output schema.
+    auto name_set = shared_expand->getAllGroupSetColumnNames();
+    // append parent_require column it may expect self-filled groupingID.
+    for (const auto & one : parent_require)
+    {
+        if (one != Expand::grouping_identifier_column_name)
+        {
+            name_set.insert(one);
+        }
+    }
+    for (const auto & grouping_name : name_set)
+    {
+        required_output.emplace_back(grouping_name);
+    }
```

Comment:
Is it equals to `expand_actions->getRequiredColumns()`?

Reply:
seems reasonable, changed